rfc/languages-in-w3s.md

```diff
@@ -0,0 +1,84 @@
+
+#### Some Notes On Our Javascript
+
+- Thankfully this team has unified on a coding style, linter, tool chain, package manager, and approach to typing. +1 on that.
+- We have not however documented why we made the choices we did. We should do so ASAP. In particular types in JSDoc style comments and why we do it needs documentation. This includes our sporadic dropping into Typescript and why we do it. (For example: 'why is there an api.js and an api.ts and the js file just says `export {}` -- the answer is 'cause modules' but is EXTREMELY not clear)
+- The w3up repo is huge, and honestly, even if you know javascript, there's a lot to get up to speed on. We should have a guide to the code.
+
+### Go
+
+- Go is a highly performant garbage collected language like Java, but is much simpler than Java
+- Go has an extremely powerful concurrency model with green threads
+- These two components together make Go one of the most powerful languages for writing servers of any kind. The challenge of server programming is not super fast compute, but super high concurrency execution. Go excels at this, which matters cause our project has a lot of servers involved.
+- Go's ability to do binary distributions by default is a major plus for kind of project.
+- By and large, people get stuff done quickly in Go because Go is extremely simple. It's very fast to learn, easy to read, and it all looks the same. The only problem is its extremely boring. (Note from Hannah: When I first started writing Go I hated it. I thought someone took the joy out of programming. Now I love it cause I get my joy from what I'm building, not my programming language)
+- Lots and lots of Filecoin tooling and software is written in Go. In fact, W3S already has Go in it, we just don't maintain it -- it's called Spade
+- Garbage collection and green threads present challenges for using Go over an FFI, or for compiling go to WebAssembly, though TinyGo has improved this story by reducing the runtime size.
+
+### Rust
+
+Caveat from Hannah: I am not a highly experienced Rust dev
+
+- Rust is a better C/C++. Like definitely much better. This is huge cause no one wrote a better C/C++ for low level programming for 20 years.
+- However, by and large, you should probably not write programs in Rust unless you'd consider writing them in C/C++. There is minimal evidence one can be as productive in Rust as higher level languages.
+- In particular, you should probably not write web servers in Rust or any other asynchronous programs unless you have extra time on your hands which we don't.
+- BTW, if you haven't programmed in C/C++ recently, you've probably forgotten that long compile times are a thing on large projects in low level languages. 
+- Rust is essential for highly performant synchronous compute tasks. That's why people use it for smart contracts, hashing, and proofs.
+- Rust has the best WebAssembly story of any language.
```

Comment:
Rust is a great tool for writing libraries that you may ever conceive running elsewhere be it native code or web, no other language offers this. I know there are some efforts to make go better there, but GC is a sticking point. Rust has no GC which is why it's good fit and has no overhead, if you have to bring your GC with you that will be a non-trivial overhead and it would need to integrate with a host GC also. Maybe one day WASM will get a GC, there is certainly an effort to make it so, but I would not make any decision that assumes it, because chances are it's not going to be ready in time.

Reply:
I 100% agree with this.

If you want your code to run absolutely anyway, Rust is the go to.